File: csrc/npu/add_sample.cpp
/**
 * Copyright (c) 2026 Huawei Technologies Co., Ltd.
 * This program is free software, you can redistribute it and/or modify it under the terms and conditions of
 * CANN Open Software License Agreement Version 2.0 (the "License").
 * Please refer to the License for details. You may not use this file except in compliance with the License.
 * THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
 * See LICENSE in the root of the software repository for the full text of the License.
 */

#include "add_sample.h"
#include "add_sample_kernel.h"

torch::Tensor add_sample(torch::Tensor src1, torch::Tensor src2)
{
    uint32_t valueNum = src1.numel();
    torch::Tensor dst = torch::zeros_like(src1);
    aclrtStream stream = nullptr;

    aclrtCreateStream(&stream);
    LaunchAddSampleKernel(src1.data_ptr<uint8_t>(), src2.data_ptr<uint8_t>(), dst.data_ptr<uint8_t>(), valueNum, stream);
    aclrtSynchronizeStream(stream);
    aclrtDestroyStream(stream);
    return dst;
}

File: python/ops_gnn/add_sample.py
"""
Copyright (c) 2026 Huawei Technologies Co., Ltd.
This program is free software, you can redistribute it and/or modify it under the terms and conditions of
CANN Open Software License Agreement Version 2.0 (the "License").
Please refer to the License for details. You may not use this file except in compliance with the License.
THIS SOFTWARE IS PROVIDED ON AN "AS IS" BASIS, WITHOUT WARRANTIES OF ANY KIND, EITHER EXPRESS OR IMPLIED,
INCLUDING BUT NOT LIMITED TO NON-INFRINGEMENT, MERCHANTABILITY, OR FITNESS FOR A PARTICULAR PURPOSE.
See LICENSE in the root of the software repository for the full text of the License.
"""

from typing import Tuple

import torch

from . import _pybind
from .typing import Tensor


def add_sample(
    src1: Tensor,
    src2: Tensor,
) -> Tensor:
    """
    Performs element-wise addition of two tensors on NPU.

    Args:
        src1 (Tensor): The first input tensor. Must be on NPU device.
        src2 (Tensor): The second input tensor. Must be on NPU device and have
            the same shape as :obj:`src1`.

    Returns:
        Tensor: The element-wise sum of :obj:`src1` and :obj:`src2`, on NPU device.

    Example:
        >>> import ops_gnn
        >>> src1 = torch.tensor([1, 2, 3], dtype=torch.uint8, device='npu')
        >>> src2 = torch.tensor([4, 5, 6], dtype=torch.uint8, device='npu')
        >>> result = ops_gnn.add_sample(src1, src2)
        >>> result
        tensor([5, 7, 9], device='npu:0', dtype=torch.uint8)
    """
    return _pybind.add_sample(src1, src2)
